src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+                // first, or maybe one of its children should be extracted earlier.
+                var topChild = this.nodes[i];
+                var childrenIndexesLimit = Math.Min(i + Arity, this.nodes.Count);
+                int topChildIndex = i;
+
+                while (++i < childrenIndexesLimit)
+                {
+                    var child = this.nodes[i];
+                    if (this.Comparer.Compare(child.priority, topChild.priority) < 0)
+                    {
+                        topChild = child;
+                        topChildIndex = i;
+                    }
+                }
+
+                // In case no child needs to be extracted earlier than the current node,
+                // there is nothing more to do - the right spot was found.
+                if (this.Comparer.Compare(node.priority, topChild.priority) <= 0)
+                {
+                    break;
+                }
+
+                // Move the top child up by one node and now investigate the
+                // node that was considered to be the top child (recursive).
+                this.PutAt(topChild, nodeIndex);
+                nodeIndex = topChildIndex;
+            }
+
+            this.PutAt(node, nodeIndex);
+        }
+
+        /// <summary>
+        /// Gets a collection that enumerates the elements of the queue.
+        /// </summary>
+        public UnorderedItemsCollection UnorderedItems { get; }
+
+        public partial class UnorderedItemsCollection : IReadOnlyCollection<(TElement element, TPriority priority)>, ICollection
+        {
+            private readonly IReadOnlyCollection<(TElement element, TPriority priority)> items;
+
+            public UnorderedItemsCollection(IReadOnlyCollection<(TElement element, TPriority priority)> items)
+            {
+                this.items = items;
+            }
+
+            public int Count => this.items.Count;
+            object ICollection.SyncRoot => this;
+            bool ICollection.IsSynchronized => false;
+
+            public void CopyTo(Array array, int index) => throw new NotImplementedException();
+
+            public struct Enumerator : IEnumerator<(TElement element, TPriority priority)>, IEnumerator
+            {
+                private readonly IEnumerator<(TElement element, TPriority priority)> enumerator;
+
+                internal Enumerator(System.Collections.Generic.IEnumerator<(TElement element, TPriority priority)> enumerator)
+                {
+                    this.enumerator = enumerator;
+                }
+
+                (TElement element, TPriority priority) IEnumerator<(TElement element, TPriority priority)>.Current
+                    => this.enumerator.Current;
+
+                object IEnumerator.Current => this.enumerator.Current;
+
+                void IDisposable.Dispose() => this.enumerator.Dispose();
+                bool IEnumerator.MoveNext() => this.enumerator.MoveNext();
```

Comment:
These should both be implemented implicitly rather than explicitly. i.e. MoveNext, Dispose, and Current should all be public.